libtelnet.c

```diff
@@ -1121,6 +1137,47 @@ static void _process(telnet_t *telnet, const char *buffer, size_t size) {
 	}
 }
 
+/* perform NVT end-of-line translation for received text */
+size_t telnet_translate_eol(telnet_t *telnet, char *buffer, size_t size,
+		int *split) {
+	size_t i = 0, l = 0;
+
+	/* no translation if in BINARY mode */
+	if (telnet->flags & TELNET_FLAG_RECEIVE_BINARY)
+		return size;
+
+	/* fix up translation if split across buffer boundary */
+	if (*split && size > 0) {
+		/* CR-NUL translates to \r which was skipped at end of previous */
+		/* buffer, while CR-LF translates to \n, so just leave it */
+		if (buffer[i] == '\0') {
+			buffer[i] = '\r';
+		}
+		++i;
+	}
+
+	for ( ; i < size-1; ++i, ++l) {
+		buffer[l] = buffer[i];
+		if (buffer[i] == '\r') {
+			if (buffer[i+1] == '\n') {
+				/* CR-LF translates to \n */
+				buffer[i++] = '\n';
```

Comment:
This line looks wrong. Should it not be modifying `buffer[l]` (lowercase L) instead of `buffer[i++]`? it looks like this would break on inputs with multiple `\r\n` sequences.